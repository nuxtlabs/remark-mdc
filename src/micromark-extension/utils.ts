import type { Effects } from 'micromark-factory-space'

// Measure the number of character codes in chunks.
// Counts tabs based on their expanded size, and CR+LF as one character.
export function sizeChunks (chunks: any[]) {
  let index = -1
  let size = 0

  while (++index < chunks.length) {
    size += typeof chunks[index] === 'string' ? chunks[index].length : 1
  }

  return size
}

export function prefixSize (events: any, type: string) {
  const tail = events[events.length - 1]
  if (!tail || tail[1].type !== type) { return 0 }
  return sizeChunks(tail[2].sliceStream(tail[1]))
}

/**
 * Calculate line indention size, line indention could be consists of multiple `linePrefix` events
 * @param events parser tokens
 * @returns line indention size
 */
export function linePrefixSize (events: any[]) {
  let size = 0
  let index = events.length - 1
  let tail = events[index]
  while (index >= 0 && tail && tail[1].type === 'linePrefix' && tail[0] === 'exit') {
    size += sizeChunks(tail[2].sliceStream(tail[1]))
    index -= 1
    tail = events[index]
  }

  return size
}

/**
 * Manage token state
 */
export const useTokenState = (tokenName: string) => {
  const token = {
    isOpen: false,
    /**
     * Enter into token, close previous open token if any
     */
    enter: (effects: Effects) => {
      const initialState = token.isOpen
      token.exit(effects)
      effects.enter(tokenName)
      token.isOpen = true

      // Revert to initial state
      return () => {
        token.isOpen = initialState
      }
    },
    /**
     * Enter into token only once, if token is already open, do nothing
     */
    enterOnce: (effects: Effects) => {
      const initialState = token.isOpen

      if (!token.isOpen) {
        effects.enter(tokenName)
        token.isOpen = true
      }

      // Revert to initial state
      return () => {
        token.isOpen = initialState
      }
    },
    /**
     * Exit from token if it is open
     */
    exit: (effects: Effects) => {
      const initialState = token.isOpen
      if (token.isOpen) {
        effects.exit(tokenName)
        token.isOpen = false
      }

      // Revert to initial state
      return () => {
        token.isOpen = initialState
      }
    }
  }
  return token
}
