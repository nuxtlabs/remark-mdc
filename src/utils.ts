export const NON_UNWRAPABLE_TYPES = [
  'componentContainerSection',
  'componentContainerDataSection',
  'containerComponent',
  'leafComponent'
]
