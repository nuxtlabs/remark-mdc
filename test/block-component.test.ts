import { describe, expect } from 'vitest'
import { runMarkdownTests } from './utils'

describe('block-component', () => {
  runMarkdownTests({
    empty: {
      markdown: '::component\n::',
      expected: '::component\n::'
    },
    text: {
      markdown: '::component\ntext\n::',
      expected: '::component\ntext\n::'
    },
    'empty-slot': {
      markdown: '::component\n#text\n::',
      expected: '::component\n#text\n::'
    },
    frontmatter: {
      markdown: '::with-frontmatter\n---\nkey: value\narray:\n  - item\n  - itemKey: value\n---\n::'
    },
    frontmatter1: {
      markdown: [
        '::with-frontmatter',
        '---',
        'key: value',
        'key2:',
        '  subkey: value',
        '  subkey2: value',
        'array:',
        '  - item',
        '  - itemKey: value',
        '---',
        '::'
      ].join('\n'),
      expected: [
        '::with-frontmatter',
        '---',
        'key: value',
        'key2.subkey: value',
        'key2.subkey2: value',
        'array:',
        '  - item',
        '  - itemKey: value',
        '---',
        '::'
      ].join('\n')
    },
    'nested-component': {
      markdown: [
        '::with-frontmatter-and-nested-component',
        '---',
        'key: value',
        'array:',
        '  - item',
        '  - itemKey: value',
        '---',
        'Default slot',
        '',
        '#secondary-slot',
        'Secondary slot value',
        '',
        '  :::hello',
        '  ---',
        '  key: value',
        '  ---',
        '  :::',
        '::'
      ].join('\n')
    },
    'section-order': {
      markdown: [
        '::comp',
        '#title',
        'Hello',
        '#another-title',
        'World',
        '#default',
        'P1',
        '',
        'P2',
        '',
        'P',
        '::'
      ].join('\n'),
      expected: [
        '::comp',
        'P1',
        '',
        'P2',
        '',
        'P',
        '',
        '#title',
        'Hello',
        '',
        '#another-title',
        'World',
        '::'
      ].join('\n')
    },
    'section-order-2': {
      markdown: [
        '::comp',
        '#a',
        'A',
        '#c',
        'C',
        '#b',
        'B',
        '#default',
        'P1',
        '::'
      ].join('\n'),
      expected: [
        '::comp',
        'P1',
        '',
        '#a',
        'A',
        '',
        '#c',
        'C',
        '',
        '#b',
        'B',
        '::'
      ].join('\n')
    },
    'ignore-code-fence': {
      markdown: [
        '::component',
        'First line',
        '',
        '```cpp',
        '#include <iostream>',
        '```',
        '',
        'Second line',
        '::',
        '',
        'Third line'
      ].join('\n')
    },
    'dangling-list': {
      markdown: [
        '::component',
        '-   list item',
        '-   list item',
        '',
        '#slot',
        'slot content',
        '::'
      ].join('\n')
    },
    'trim-slot-name': {
      markdown: [
        '::component',
        '#slot ',
        'slot content',
        '::'
      ].join('\n'),
      expected: [
        '::component',
        '#slot',
        'slot content',
        '::'
      ].join('\n')
    },
    'sugar-syntax': {
      markdown: [
        ':component'
      ].join('\n'),
      extra: (_md, ast) => {
        expect(ast.children[0].type).toBe('textComponent')
        expect(ast.children[0].name).toBe('component')
      }
    }
  })
})
