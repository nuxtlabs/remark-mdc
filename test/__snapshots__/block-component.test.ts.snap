// Vitest Snapshot v1

exports[`block-component > danglig-list 1`] = `
{
  "children": [
    {
      "attributes": {},
      "children": [
        {
          "children": [
            {
              "checked": null,
              "children": [
                {
                  "children": [
                    {
                      "position": {
                        "end": {
                          "column": 14,
                          "line": 2,
                          "offset": 25,
                        },
                        "start": {
                          "column": 5,
                          "line": 2,
                          "offset": 16,
                        },
                      },
                      "type": "text",
                      "value": "list item",
                    },
                  ],
                  "position": {
                    "end": {
                      "column": 14,
                      "line": 2,
                      "offset": 25,
                    },
                    "start": {
                      "column": 5,
                      "line": 2,
                      "offset": 16,
                    },
                  },
                  "type": "paragraph",
                },
              ],
              "position": {
                "end": {
                  "column": 14,
                  "line": 2,
                  "offset": 25,
                },
                "start": {
                  "column": 1,
                  "line": 2,
                  "offset": 12,
                },
              },
              "spread": false,
              "type": "listItem",
            },
            {
              "checked": null,
              "children": [
                {
                  "children": [
                    {
                      "position": {
                        "end": {
                          "column": 14,
                          "line": 3,
                          "offset": 39,
                        },
                        "start": {
                          "column": 5,
                          "line": 3,
                          "offset": 30,
                        },
                      },
                      "type": "text",
                      "value": "list item",
                    },
                  ],
                  "position": {
                    "end": {
                      "column": 14,
                      "line": 3,
                      "offset": 39,
                    },
                    "start": {
                      "column": 5,
                      "line": 3,
                      "offset": 30,
                    },
                  },
                  "type": "paragraph",
                },
              ],
              "position": {
                "end": {
                  "column": 6,
                  "line": 5,
                  "offset": 46,
                },
                "start": {
                  "column": 1,
                  "line": 3,
                  "offset": 26,
                },
              },
              "spread": true,
              "type": "listItem",
            },
          ],
          "ordered": false,
          "position": {
            "end": {
              "column": 6,
              "line": 5,
              "offset": 46,
            },
            "start": {
              "column": 1,
              "line": 2,
              "offset": 12,
            },
          },
          "spread": false,
          "start": null,
          "type": "list",
        },
        {
          "attributes": {},
          "children": [
            {
              "children": [
                {
                  "position": {
                    "end": {
                      "column": 13,
                      "line": 6,
                      "offset": 59,
                    },
                    "start": {
                      "column": 1,
                      "line": 6,
                      "offset": 47,
                    },
                  },
                  "type": "text",
                  "value": "slot content",
                },
              ],
              "position": {
                "end": {
                  "column": 13,
                  "line": 6,
                  "offset": 59,
                },
                "start": {
                  "column": 1,
                  "line": 6,
                  "offset": 47,
                },
              },
              "type": "paragraph",
            },
          ],
          "data": {
            "hName": "component-slot",
            "hProperties": {
              "v-slot:slot": "",
            },
          },
          "name": "slot",
          "position": {
            "end": {
              "column": 3,
              "line": 7,
              "offset": 62,
            },
            "start": {
              "column": 2,
              "line": 5,
              "offset": 42,
            },
          },
          "type": "componentContainerSection",
        },
      ],
      "data": {
        "hName": "component",
        "hProperties": {},
      },
      "fmAttributes": {},
      "name": "component",
      "position": {
        "end": {
          "column": 3,
          "line": 7,
          "offset": 62,
        },
        "start": {
          "column": 1,
          "line": 1,
          "offset": 0,
        },
      },
      "rawData": undefined,
      "type": "containerComponent",
    },
  ],
  "position": {
    "end": {
      "column": 3,
      "line": 7,
      "offset": 62,
    },
    "start": {
      "column": 1,
      "line": 1,
      "offset": 0,
    },
  },
  "type": "root",
}
`;

exports[`block-component > empty 1`] = `
{
  "children": [
    {
      "attributes": {},
      "children": [],
      "data": {
        "hName": "component",
        "hProperties": {},
      },
      "fmAttributes": {},
      "name": "component",
      "position": {
        "end": {
          "column": 3,
          "line": 2,
          "offset": 14,
        },
        "start": {
          "column": 1,
          "line": 1,
          "offset": 0,
        },
      },
      "type": "containerComponent",
    },
  ],
  "position": {
    "end": {
      "column": 3,
      "line": 2,
      "offset": 14,
    },
    "start": {
      "column": 1,
      "line": 1,
      "offset": 0,
    },
  },
  "type": "root",
}
`;

exports[`block-component > empty-slot 1`] = `
{
  "children": [
    {
      "attributes": {},
      "children": [
        {
          "attributes": {},
          "children": [],
          "data": {
            "hName": "component-slot",
            "hProperties": {
              "v-slot:text": "",
            },
          },
          "name": "text",
          "position": {
            "end": {
              "column": 3,
              "line": 3,
              "offset": 20,
            },
            "start": {
              "column": 2,
              "line": 2,
              "offset": 13,
            },
          },
          "type": "componentContainerSection",
        },
      ],
      "data": {
        "hName": "component",
        "hProperties": {},
      },
      "fmAttributes": {},
      "name": "component",
      "position": {
        "end": {
          "column": 3,
          "line": 3,
          "offset": 20,
        },
        "start": {
          "column": 1,
          "line": 1,
          "offset": 0,
        },
      },
      "type": "containerComponent",
    },
  ],
  "position": {
    "end": {
      "column": 3,
      "line": 3,
      "offset": 20,
    },
    "start": {
      "column": 1,
      "line": 1,
      "offset": 0,
    },
  },
  "type": "root",
}
`;

exports[`block-component > frontmatter 1`] = `
{
  "children": [
    {
      "attributes": {},
      "children": [],
      "data": {
        "hName": "with-frontmatter",
        "hProperties": {
          ":array": "[\\"item\\",{\\"itemKey\\":\\"value\\"}]",
          "key": "value",
        },
      },
      "fmAttributes": {
        "array": [
          "item",
          {
            "itemKey": "value",
          },
        ],
        "key": "value",
      },
      "name": "with-frontmatter",
      "position": {
        "end": {
          "column": 3,
          "line": 8,
          "offset": 75,
        },
        "start": {
          "column": 1,
          "line": 1,
          "offset": 0,
        },
      },
      "rawData": "
key: value
array:
  - item
  - itemKey: value
---",
      "type": "containerComponent",
    },
  ],
  "position": {
    "end": {
      "column": 3,
      "line": 8,
      "offset": 75,
    },
    "start": {
      "column": 1,
      "line": 1,
      "offset": 0,
    },
  },
  "type": "root",
}
`;

exports[`block-component > frontmatter1 1`] = `
{
  "children": [
    {
      "attributes": {},
      "children": [],
      "data": {
        "hName": "with-frontmatter",
        "hProperties": {
          ":array": "[\\"item\\",{\\"itemKey\\":\\"value\\"}]",
          ":key2": "{\\"subkey\\":\\"value\\",\\"subkey2\\":\\"value\\"}",
          "key": "value",
        },
      },
      "fmAttributes": {
        "array": [
          "item",
          {
            "itemKey": "value",
          },
        ],
        "key": "value",
        "key2": {
          "subkey": "value",
          "subkey2": "value",
        },
      },
      "name": "with-frontmatter",
      "position": {
        "end": {
          "column": 3,
          "line": 11,
          "offset": 114,
        },
        "start": {
          "column": 1,
          "line": 1,
          "offset": 0,
        },
      },
      "rawData": "
key: value
key2:
  subkey: value
  subkey2: value
array:
  - item
  - itemKey: value
---",
      "type": "containerComponent",
    },
  ],
  "position": {
    "end": {
      "column": 3,
      "line": 11,
      "offset": 114,
    },
    "start": {
      "column": 1,
      "line": 1,
      "offset": 0,
    },
  },
  "type": "root",
}
`;

exports[`block-component > ignore-code-fence 1`] = `
{
  "children": [
    {
      "attributes": {},
      "children": [
        {
          "children": [
            {
              "position": {
                "end": {
                  "column": 11,
                  "line": 2,
                  "offset": 22,
                },
                "start": {
                  "column": 1,
                  "line": 2,
                  "offset": 12,
                },
              },
              "type": "text",
              "value": "First line",
            },
          ],
          "position": {
            "end": {
              "column": 11,
              "line": 2,
              "offset": 22,
            },
            "start": {
              "column": 1,
              "line": 2,
              "offset": 12,
            },
          },
          "type": "paragraph",
        },
        {
          "lang": "cpp",
          "meta": null,
          "position": {
            "end": {
              "column": 4,
              "line": 6,
              "offset": 54,
            },
            "start": {
              "column": 1,
              "line": 4,
              "offset": 24,
            },
          },
          "type": "code",
          "value": "#include <iostream>",
        },
        {
          "children": [
            {
              "position": {
                "end": {
                  "column": 12,
                  "line": 8,
                  "offset": 67,
                },
                "start": {
                  "column": 1,
                  "line": 8,
                  "offset": 56,
                },
              },
              "type": "text",
              "value": "Second line",
            },
          ],
          "position": {
            "end": {
              "column": 12,
              "line": 8,
              "offset": 67,
            },
            "start": {
              "column": 1,
              "line": 8,
              "offset": 56,
            },
          },
          "type": "paragraph",
        },
      ],
      "data": {
        "hName": "component",
        "hProperties": {},
      },
      "fmAttributes": {},
      "name": "component",
      "position": {
        "end": {
          "column": 3,
          "line": 9,
          "offset": 70,
        },
        "start": {
          "column": 1,
          "line": 1,
          "offset": 0,
        },
      },
      "type": "containerComponent",
    },
    {
      "children": [
        {
          "position": {
            "end": {
              "column": 11,
              "line": 11,
              "offset": 82,
            },
            "start": {
              "column": 1,
              "line": 11,
              "offset": 72,
            },
          },
          "type": "text",
          "value": "Third line",
        },
      ],
      "position": {
        "end": {
          "column": 11,
          "line": 11,
          "offset": 82,
        },
        "start": {
          "column": 1,
          "line": 11,
          "offset": 72,
        },
      },
      "type": "paragraph",
    },
  ],
  "position": {
    "end": {
      "column": 11,
      "line": 11,
      "offset": 82,
    },
    "start": {
      "column": 1,
      "line": 1,
      "offset": 0,
    },
  },
  "type": "root",
}
`;

exports[`block-component > nested-component 1`] = `
{
  "children": [
    {
      "attributes": {},
      "children": [
        {
          "children": [
            {
              "position": {
                "end": {
                  "column": 13,
                  "line": 8,
                  "offset": 106,
                },
                "start": {
                  "column": 1,
                  "line": 8,
                  "offset": 94,
                },
              },
              "type": "text",
              "value": "Default slot",
            },
          ],
          "position": {
            "end": {
              "column": 13,
              "line": 8,
              "offset": 106,
            },
            "start": {
              "column": 1,
              "line": 8,
              "offset": 94,
            },
          },
          "type": "paragraph",
        },
        {
          "attributes": {},
          "children": [
            {
              "children": [
                {
                  "position": {
                    "end": {
                      "column": 21,
                      "line": 11,
                      "offset": 144,
                    },
                    "start": {
                      "column": 1,
                      "line": 11,
                      "offset": 124,
                    },
                  },
                  "type": "text",
                  "value": "Secondary slot value",
                },
              ],
              "position": {
                "end": {
                  "column": 21,
                  "line": 11,
                  "offset": 144,
                },
                "start": {
                  "column": 1,
                  "line": 11,
                  "offset": 124,
                },
              },
              "type": "paragraph",
            },
            {
              "attributes": {},
              "children": [],
              "data": {
                "hName": "hello",
                "hProperties": {
                  "key": "value",
                },
              },
              "fmAttributes": {
                "key": "value",
              },
              "name": "hello",
              "position": {
                "end": {
                  "column": 6,
                  "line": 17,
                  "offset": 187,
                },
                "start": {
                  "column": 3,
                  "line": 13,
                  "offset": 148,
                },
              },
              "rawData": "
  key: value
  ---",
              "type": "containerComponent",
            },
          ],
          "data": {
            "hName": "component-slot",
            "hProperties": {
              "v-slot:secondary-slot": "",
            },
          },
          "name": "secondary-slot",
          "position": {
            "end": {
              "column": 3,
              "line": 18,
              "offset": 190,
            },
            "start": {
              "column": 2,
              "line": 10,
              "offset": 109,
            },
          },
          "type": "componentContainerSection",
        },
      ],
      "data": {
        "hName": "with-frontmatter-and-nested-component",
        "hProperties": {
          ":array": "[\\"item\\",{\\"itemKey\\":\\"value\\"}]",
          "key": "value",
        },
      },
      "fmAttributes": {
        "array": [
          "item",
          {
            "itemKey": "value",
          },
        ],
        "key": "value",
      },
      "name": "with-frontmatter-and-nested-component",
      "position": {
        "end": {
          "column": 3,
          "line": 18,
          "offset": 190,
        },
        "start": {
          "column": 1,
          "line": 1,
          "offset": 0,
        },
      },
      "rawData": "
key: value
array:
  - item
  - itemKey: value
---",
      "type": "containerComponent",
    },
  ],
  "position": {
    "end": {
      "column": 3,
      "line": 18,
      "offset": 190,
    },
    "start": {
      "column": 1,
      "line": 1,
      "offset": 0,
    },
  },
  "type": "root",
}
`;

exports[`block-component > section-order 1`] = `
{
  "children": [
    {
      "attributes": {},
      "children": [
        {
          "attributes": {},
          "children": [
            {
              "children": [
                {
                  "position": {
                    "end": {
                      "column": 6,
                      "line": 3,
                      "offset": 19,
                    },
                    "start": {
                      "column": 1,
                      "line": 3,
                      "offset": 14,
                    },
                  },
                  "type": "text",
                  "value": "Hello",
                },
              ],
              "position": {
                "end": {
                  "column": 6,
                  "line": 3,
                  "offset": 19,
                },
                "start": {
                  "column": 1,
                  "line": 3,
                  "offset": 14,
                },
              },
              "type": "paragraph",
            },
          ],
          "data": {
            "hName": "component-slot",
            "hProperties": {
              "v-slot:title": "",
            },
          },
          "name": "title",
          "position": {
            "end": {
              "column": 1,
              "line": 4,
              "offset": 20,
            },
            "start": {
              "column": 2,
              "line": 2,
              "offset": 8,
            },
          },
          "type": "componentContainerSection",
        },
        {
          "attributes": {},
          "children": [
            {
              "children": [
                {
                  "position": {
                    "end": {
                      "column": 6,
                      "line": 5,
                      "offset": 40,
                    },
                    "start": {
                      "column": 1,
                      "line": 5,
                      "offset": 35,
                    },
                  },
                  "type": "text",
                  "value": "World",
                },
              ],
              "position": {
                "end": {
                  "column": 6,
                  "line": 5,
                  "offset": 40,
                },
                "start": {
                  "column": 1,
                  "line": 5,
                  "offset": 35,
                },
              },
              "type": "paragraph",
            },
          ],
          "data": {
            "hName": "component-slot",
            "hProperties": {
              "v-slot:another-title": "",
            },
          },
          "name": "another-title",
          "position": {
            "end": {
              "column": 1,
              "line": 6,
              "offset": 41,
            },
            "start": {
              "column": 2,
              "line": 4,
              "offset": 21,
            },
          },
          "type": "componentContainerSection",
        },
        {
          "children": [
            {
              "position": {
                "end": {
                  "column": 3,
                  "line": 7,
                  "offset": 52,
                },
                "start": {
                  "column": 1,
                  "line": 7,
                  "offset": 50,
                },
              },
              "type": "text",
              "value": "P1",
            },
          ],
          "position": {
            "end": {
              "column": 3,
              "line": 7,
              "offset": 52,
            },
            "start": {
              "column": 1,
              "line": 7,
              "offset": 50,
            },
          },
          "type": "paragraph",
        },
        {
          "children": [
            {
              "position": {
                "end": {
                  "column": 3,
                  "line": 9,
                  "offset": 56,
                },
                "start": {
                  "column": 1,
                  "line": 9,
                  "offset": 54,
                },
              },
              "type": "text",
              "value": "P2",
            },
          ],
          "position": {
            "end": {
              "column": 3,
              "line": 9,
              "offset": 56,
            },
            "start": {
              "column": 1,
              "line": 9,
              "offset": 54,
            },
          },
          "type": "paragraph",
        },
        {
          "children": [
            {
              "position": {
                "end": {
                  "column": 2,
                  "line": 11,
                  "offset": 59,
                },
                "start": {
                  "column": 1,
                  "line": 11,
                  "offset": 58,
                },
              },
              "type": "text",
              "value": "P",
            },
          ],
          "position": {
            "end": {
              "column": 2,
              "line": 11,
              "offset": 59,
            },
            "start": {
              "column": 1,
              "line": 11,
              "offset": 58,
            },
          },
          "type": "paragraph",
        },
      ],
      "data": {
        "hName": "comp",
        "hProperties": {},
      },
      "fmAttributes": {},
      "name": "comp",
      "position": {
        "end": {
          "column": 3,
          "line": 12,
          "offset": 62,
        },
        "start": {
          "column": 1,
          "line": 1,
          "offset": 0,
        },
      },
      "rawData": undefined,
      "type": "containerComponent",
    },
  ],
  "position": {
    "end": {
      "column": 3,
      "line": 12,
      "offset": 62,
    },
    "start": {
      "column": 1,
      "line": 1,
      "offset": 0,
    },
  },
  "type": "root",
}
`;

exports[`block-component > section-order-2 1`] = `
{
  "children": [
    {
      "attributes": {},
      "children": [
        {
          "attributes": {},
          "children": [
            {
              "children": [
                {
                  "position": {
                    "end": {
                      "column": 2,
                      "line": 3,
                      "offset": 11,
                    },
                    "start": {
                      "column": 1,
                      "line": 3,
                      "offset": 10,
                    },
                  },
                  "type": "text",
                  "value": "A",
                },
              ],
              "position": {
                "end": {
                  "column": 2,
                  "line": 3,
                  "offset": 11,
                },
                "start": {
                  "column": 1,
                  "line": 3,
                  "offset": 10,
                },
              },
              "type": "paragraph",
            },
          ],
          "data": {
            "hName": "component-slot",
            "hProperties": {
              "v-slot:a": "",
            },
          },
          "name": "a",
          "position": {
            "end": {
              "column": 1,
              "line": 4,
              "offset": 12,
            },
            "start": {
              "column": 2,
              "line": 2,
              "offset": 8,
            },
          },
          "type": "componentContainerSection",
        },
        {
          "attributes": {},
          "children": [
            {
              "children": [
                {
                  "position": {
                    "end": {
                      "column": 2,
                      "line": 5,
                      "offset": 16,
                    },
                    "start": {
                      "column": 1,
                      "line": 5,
                      "offset": 15,
                    },
                  },
                  "type": "text",
                  "value": "C",
                },
              ],
              "position": {
                "end": {
                  "column": 2,
                  "line": 5,
                  "offset": 16,
                },
                "start": {
                  "column": 1,
                  "line": 5,
                  "offset": 15,
                },
              },
              "type": "paragraph",
            },
          ],
          "data": {
            "hName": "component-slot",
            "hProperties": {
              "v-slot:c": "",
            },
          },
          "name": "c",
          "position": {
            "end": {
              "column": 1,
              "line": 6,
              "offset": 17,
            },
            "start": {
              "column": 2,
              "line": 4,
              "offset": 13,
            },
          },
          "type": "componentContainerSection",
        },
        {
          "attributes": {},
          "children": [
            {
              "children": [
                {
                  "position": {
                    "end": {
                      "column": 2,
                      "line": 7,
                      "offset": 21,
                    },
                    "start": {
                      "column": 1,
                      "line": 7,
                      "offset": 20,
                    },
                  },
                  "type": "text",
                  "value": "B",
                },
              ],
              "position": {
                "end": {
                  "column": 2,
                  "line": 7,
                  "offset": 21,
                },
                "start": {
                  "column": 1,
                  "line": 7,
                  "offset": 20,
                },
              },
              "type": "paragraph",
            },
          ],
          "data": {
            "hName": "component-slot",
            "hProperties": {
              "v-slot:b": "",
            },
          },
          "name": "b",
          "position": {
            "end": {
              "column": 1,
              "line": 8,
              "offset": 22,
            },
            "start": {
              "column": 2,
              "line": 6,
              "offset": 18,
            },
          },
          "type": "componentContainerSection",
        },
        {
          "children": [
            {
              "position": {
                "end": {
                  "column": 3,
                  "line": 9,
                  "offset": 33,
                },
                "start": {
                  "column": 1,
                  "line": 9,
                  "offset": 31,
                },
              },
              "type": "text",
              "value": "P1",
            },
          ],
          "position": {
            "end": {
              "column": 3,
              "line": 9,
              "offset": 33,
            },
            "start": {
              "column": 1,
              "line": 9,
              "offset": 31,
            },
          },
          "type": "paragraph",
        },
      ],
      "data": {
        "hName": "comp",
        "hProperties": {},
      },
      "fmAttributes": {},
      "name": "comp",
      "position": {
        "end": {
          "column": 3,
          "line": 10,
          "offset": 36,
        },
        "start": {
          "column": 1,
          "line": 1,
          "offset": 0,
        },
      },
      "rawData": undefined,
      "type": "containerComponent",
    },
  ],
  "position": {
    "end": {
      "column": 3,
      "line": 10,
      "offset": 36,
    },
    "start": {
      "column": 1,
      "line": 1,
      "offset": 0,
    },
  },
  "type": "root",
}
`;

exports[`block-component > text 1`] = `
{
  "children": [
    {
      "attributes": {},
      "children": [
        {
          "children": [
            {
              "position": {
                "end": {
                  "column": 5,
                  "line": 2,
                  "offset": 16,
                },
                "start": {
                  "column": 1,
                  "line": 2,
                  "offset": 12,
                },
              },
              "type": "text",
              "value": "text",
            },
          ],
          "position": {
            "end": {
              "column": 5,
              "line": 2,
              "offset": 16,
            },
            "start": {
              "column": 1,
              "line": 2,
              "offset": 12,
            },
          },
          "type": "paragraph",
        },
      ],
      "data": {
        "hName": "component",
        "hProperties": {},
      },
      "fmAttributes": {},
      "name": "component",
      "position": {
        "end": {
          "column": 3,
          "line": 3,
          "offset": 19,
        },
        "start": {
          "column": 1,
          "line": 1,
          "offset": 0,
        },
      },
      "type": "containerComponent",
    },
  ],
  "position": {
    "end": {
      "column": 3,
      "line": 3,
      "offset": 19,
    },
    "start": {
      "column": 1,
      "line": 1,
      "offset": 0,
    },
  },
  "type": "root",
}
`;
